### ISSUE TITLE
"load.go" is failing in large clusters

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#20638]
"load.go" is currently regularly failing in 1000-node clusters

Generally it is failing like that:

      creating rc load-test-small-rc-1703
      Expected error:
          <*errors.errorString | 0xc20e439b00>: {
              s: "Only 0 pods started out of 5",
          }
          Only 0 pods started out of 5
      not to have occurred

@gmarek @davidopp

**Comment by wojtek-t** [comment]
I started some debugging around that today, and found something interesting.

Basically it seems that there is a huge time difference between when we send request to apiserver and when apiserver actually observes it.

To give more details, I added some logs in the reflector.go as the following:

    229 func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
    ....
    238   glog.Errorf("XXX %s: listing", r.name)
    239   list, err := r.listerWatcher.List(options)
    240   if err != nil {
    241     return fmt.Errorf("%s: Failed to list %v: %v", r.name, r.expectedType, err)
    242   }
    243   glog.Errorf("XXX %s: acceessor", r.name)
    244   metaInterface, err := meta.Accessor(list)

And the logs for example replication controller from the test are the following:

    E0204 14:33:32.021700   10205 reflector.go:238] XXX load-test-small-rc-539: listing
    ...
    E0204 14:33:42.417496   10205 reflector.go:243] XXX load-test-small-rc-539: acceessor

and from apiserver logs:

    I0204 13:33:42.348435    3503 handlers.go:130] GET /api/v1/namespaces/e2e-tests-load-1kyv5/pods?labelSelector=name%3Dload-test-small-rc-539&resourceVersion=0: (1.829487ms) 200 [[e2e.test/v1.2.0 (linux/amd64) kubernetes/ec7d298] 74.125.60.194:48260]


So basically, the request is very quickly processed by apiserver, but it is __hanging for ~10 seconds__ somewhere between the client and apiserver.

Note, that I've seen a difference between "listing" and "accessor" above for other RCs exceeding even 5 minutes, which is completely ridiculous

@lavalamp @krousey - any thoughts about it ^

**Comment by wojtek-t** [comment]
As an experiment I tried increasing MaxIdleConnsPerHost to 1000, but it didn't change anything...

**Comment by wojtek-t** [comment]
@kubernetes/sig-scalability

**Comment by lavalamp** [comment]
Did you measure your clock skew?

If the traffic went from within the cluster, did it appear in kube-proxy's log?

**Comment by lavalamp** [comment]
Also did it have to retry the request? I think apiserver can ask clients to try again later if it's busy (maybe I'm misremembering a planned feature).

**Comment by wojtek-t** [comment]
It's definitely not clock skew - the clock skew is negligible.

I though about retries, but IIUC, if apiserver rejects a request with 429 error, the it should be visible in its logs. But it's nothing like that there...

The trafic is not from inside the cluster - the call is from the test itself, and the test was running on my machine. And answering the question that you may have - it doesn't seem to be related to the distance from my machine to the apiserver - the answer from apiserver is coming pretty fast.

**Comment by lavalamp** [comment]
Is apiserver so heavily loaded that it just takes that long to accept the
connection?

**Comment by wojtek-t** [comment]
It doesn't seem to be (I mean it's not cpu-bounded at least)

**Comment by wojtek-t** [comment]
For me it looks more like some problem at the test side - like test cannot set up more connections, or is creating them slowly, or sth like that...

Maybe the problem is that there is just a single client in e2e test framework an all requests to apiserver are send from a single client (and there are potentially O(thousands) lists/watches that can co-exist at the same time)?

**Comment by wojtek-t** [comment]
OK - I have a confirmation that the problem is somewhere in the test or client.

I added additional log in pkg/client/unversioned/request.go, just before where Do() is called:

     780     time.Sleep(r.backoffMgr.CalculateBackoff(r.URL()))
     784     glog.Errorf("calling: %s", url)
     785     resp, err := client.Do(req)

With those logs I have the logs as following:

    E0205 12:59:05.909913   29557 reflector.go:238] XXX load-test-small-rc-2886: listing
    ...
    E0205 12:59:24.360847   29557 request.go:784] calling: https://104.197.56.151/api/v1/namespaces/e2e-tests-load-54y0g/pods?labelSelector=name%3Dload-test-small-rc-2886&resourceVersion=0
    ...
    E0205 12:59:24.881035   29557 reflector.go:243] XXX load-test-small-rc-2886: acceessor 

and logs from apiserver:

    I0205 12:59:24.804454   3495 handlers.go:130] GET api/v1/namespaces/e2e-tests-load-54y0g/pods?labelSelector=name%3Dload-test-small-rc-2886&resourceVersion=0: (3.57838ms) 200 [[e2e.test/v1.2.0 (linux/amd64) kubernetes/99e56c8] 74.125.60.194:38005]

 which in my opinion shows that the problem is somewhere even before we really trying to send the http request...

**Comment by wojtek-t** [comment]
And I verified that it's not a problem of backoff...

**Comment by wojtek-t** [comment]
heh... - it seems that it might be rate-limitting at the client level
I'm currently verifying it...

**Comment by wojtek-t** [comment]
Yes - increasing QPS limit solved that particular issue. There seem to be some other, but I will continue debugging them.

**Comment by lavalamp** [comment]
We should be careful about increasing the test client's rate limit-- we
don't want to make so many requests that we distort the performance numbers
we get.

On Fri, Feb 5, 2016 at 6:33 AM, Wojciech Tyczynski <notifications@github.com

**Comment by wojtek-t** [comment]
@lavalamp - I agree that we shouldn't increase it in general.
However, load test is very specific, so that we do a bunch of read-only operations and there is only one such test client, so I think it's pretty safe to change it. Note that I'm changing limits only for the client used in this particular test, so this will not affect any other tests.
Additionally, if we increase the limits, it can only make things worse for the system (so it may give us false positives).

Anyway - the problem here was basically that if we were starting watch we called List, and then when we were calling watch, it was throttled and called some time later, which resulted in "error too old", and caused to call another list. We were accumulating those calls over time, which resulted in system making pretty good progress, but test itself wasn't really following the current state. Because those were mainly read-only operations, it's pretty safe to increase the limits for this particular test imho.

**Comment by wojtek-t** [comment]
To clarify - I already did some tests today, and it solves that particular problem. I will try to finish investigation on Monday (but I think I understand what other problems we have there, and those seem to be test-specific things).

**Comment by lavalamp** [comment]
Changing this test only SGTM. Thanks for bringing that failure mode to my
attention, I will file an issue because we need to fix that everywhere.

**Comment by wojtek-t** [comment]
I added some more details there.

Also - as a result of that, I'm planning to add a metric of "re-lists" and once we have it I will think if we can minimize them (I have some idea, but need to verify if it's worth doing).

**Comment by wojtek-t** [comment]
After few fixes, I've enabled load.go test in 500-node kubemarks and since enabling (6 runs), it is passing.
I've just enabled it in 1000-node kubemarks and we will see what will happen.

**Comment by wojtek-t** [comment]
The test is failing in 1000-node clusters.
At first, I see a bunch of logs like that:

    10:19:32 Feb 11 10:19:32.824: INFO: Unexpected error occurred: Number of reported pods changed: 10 vs 5

**Comment by wojtek-t** [comment]
Also, there listing replication controllers is pretty long. But looking into traces even from much smaller clusters they seem to look like:

    I0212 13:07:15.732252    2961 trace.go:57] Trace "List /api/v1/namespaces/e2e-tests-load-urk9t/replicationcontrollers" (started 2016-02-12 13:07:15.179169119 +0000 UTC):
    [4.646µs] [4.646µs] About to List from storage
    [13.372281ms] [13.367635ms] Listing from storage done
    [14.622092ms] [1.249811ms] Self-linking done
    [553.031529ms] [538.409437ms] Writing http response done (328 items)
    [553.033668ms] [2.139µs] END

I'm wondering if there is anything we can with "writing http response":
https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/resthandler.go#L325

**Comment by wojtek-t** [comment]
And what is more interesting, is that Kubemark-500 generally works and listing replication controllers there is faster than in Kubemark-100, which is extremely strange...

**Comment by wojtek-t** [comment]
Regarding why load.go test is not working correctly on 100-node Kubemark cluster.
I've added a bunch of different traces and it clearly seems like some network issues.

In particular, I've added some traces in json Serializer EncodeToStream method:
https://github.com/kubernetes/kubernetes/blob/master/pkg/runtime/serializer/json/json.go#L158

What I did is I changed the end of the function to the following:

    187   trace := util.NewTrace("json encode " + reflect.TypeOf(obj).String())
    188   defer trace.LogIfLong(250 * time.Millisecond)
    189   buf := &bytes.Buffer{}
    190   encoder2 := json.NewEncoder(buf)
    191   if err := encoder2.Encode(obj); err != nil {
    192     glog.Errorf("!!!!! ERROR: %v", err)
    193   }
    194   trace.Step("object encoded")
    195 
    196   encoder := json.NewEncoder(w)
    197   return encoder.Encode(obj)

What I got, is a huge number of traces like that:

    I0216 15:30:29.466957  2923 trace.go:57] Trace "json encode *v1.ReplicationControllerList" (started 2016-02-16 15:30:29.055674269 +0000 UTC):
    [6.283276ms] [6.283276ms] object encoded
    [411.23634ms] [404.951964ms] END

This means that encoding the object itself is pretty fast (<10ms), but encoding + writing it the the http response writer takes 400ms+

I still don't really understand what is causing it.

[BTW, I've tried using bigger master machine to get higher network throughput available, but it doesn't change anything - so it seems that we are not throttled at the GCE networking layer. It must be something different.]

**Comment by timothysc** [comment]
@wojtek-t how many sockets do you have open.  When we did our recent tests at 100 we saw a very large number of sockets open.

**Comment by wojtek-t** [comment]
The mystery thing is that we are running 500-node kubemarks which are better (although everything is just 5x bigger).

**Comment by timothysc** [comment]
If that's the case then there is likely a lock on a resource.

**Comment by wojtek-t** [comment]
Or some configuration issue....

**Comment by wojtek-t** [comment]
It seems that I understand it. Basically the problem is that we are running the kubemark cluster far from Jenkins master (or my local machine when I'm doing it manually). I will just change the location where we are running it.

**Comment by wojtek-t** [comment]
Yup - I've move Kubemark closer to Jenkins master and now it works ok.

So to summarize:
- we have working load.go test on 100- and 500-node Kubemark
- the remaining thing is to figure our failure on 1000-node Kubemark

Fortunately, now we will have logs from master components thanks to #21307 so it should be easier on the next failure. I will start looking deeply into it tomorrow (when we have logs for the first time)

**Comment by davidopp** [comment]
Moving to v1.2 P2 to indicate nice-to-have for 1.2 but not blocker.

**Comment by wojtek-t** [comment]
I managed to debug something serious here - see: #21672

**Comment by wojtek-t** [comment]
To be honest, I think that #21672 is the only serious problem there.

**Comment by wojtek-t** [comment]
Yeah - and we managed to run load test successfully in 1000-node kubemark. So all problems seem to be solved now.

**Comment by davidopp** [comment]
@wojtek-t please close this issue if you think it's finished.

**Comment by wojtek-t** [comment]
We have success rate of the test <50%, so I will keep it open. But this seems more like test issues (or stability issues) than performance issue. However, I will try to look into it a bit more.

**Comment by davidopp** [comment]
Moving out of v1.2 milestone, upgrading to P1. (Feel free to downgrade if you think it's not P1-worthy, @wojtek-t).

**Comment by wojtek-t** [comment]
Actually it seems to be fixed now.
